audit/audit.go

```diff
@@ -0,0 +1,130 @@
+package audit
+
+//this can't import kafka, as kafka already imports log
+//this either has to go in kafka or a separate package
+
+import (
+	"context"
+	"encoding/json"
+	"github.com/ONSdigital/go-ns/identity"
+	"github.com/ONSdigital/go-ns/log"
+	"net/http"
+	"time"
+)
+
+type contextKey string
+
+const eventContextKey = contextKey("audit")
+const fromContextKey = contextKey("from")
+const requestIDHeader = "X-Request-Id"
+const fromHeader = "from"
+
+type Event struct {
+	Created         string         `avro:"created"`
+	Service         string         `avro:"service"`
+	Namespace       string         `avro:"name_space"`
+	RequestID       string         `avro:"request_id"`
+	User            string         `avro:"user"`
+	AttemptedAction string         `avro:"attempted_action"`
+	Result          string         `avro:"result"`
+	Params          []keyValuePair `avro:"params"`
+}
+
+type keyValuePair struct {
+	Key   string `avro:"key"`
+	Value string `avro:"value"`
+}
+
+type OutboundProducer interface {
+	Output() chan []byte
+}
+
+type Auditor struct {
+	Namespace string
+	TokenName string
+	Producer  OutboundProducer
+}
+
+func New(namespace string, producer OutboundProducer, token string) *Auditor {
+	return &Auditor{
+		Namespace: namespace,
+		Producer:  producer,
+		TokenName: token,
+	}
+}
+
+func (a *Auditor) Interceptor() func(handler http.Handler) http.Handler {
+	return func(handler http.Handler) http.Handler {
+		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			ctx := r.Context()
+			event := &Event{
+				Params: make([]keyValuePair, 0),
+			}
+
+			if serviceToken := r.Header.Get(a.TokenName); len(serviceToken) > 0 {
+				ctx = context.WithValue(ctx, a.TokenName, serviceToken)
+				event.Service = serviceToken
+			}
+
+			if from := r.Header.Get(fromHeader); len(from) > 0 {
+				ctx = context.WithValue(ctx, fromContextKey, from)
+				event.User = from
+			}
+
+			event.RequestID = r.Header.Get(requestIDHeader)
+			event.Namespace = a.Namespace
+
+			eventString, err := json.Marshal(event)
+			if err != nil {
+				//we couldn't create the audit event. some input must be invalid
```

Comment:
This shouldn't happen if we set json tags against the event struct, where the field can be empty set `omitempty` (see below), also I would agree if there is an error here we would want to logout the error and force an internal server error in the response, `w http.ResponseWriter`

```
type Event struct {
    Created          string                `avro:"created"            json:"created,omitempty"`
    Service          string                `avro:"service"            json:"service"`
    Namespace        string                `avro:"name_space"         json:"name_space"`
    RequestID        string                `avro:"request_id"         json:"request_id"`
    User             string                `avro:"user"               json:"user"`
    AttemptedAction  string                `avro:"attempted_action"   json:"attempted_action,omitempty"`
    Result           string                `avro:"result"             json:"result,omitempty"`
    Params           []keyValuePair        `avro:"params"             json:"params"`
}

type keyValuePair struct {
    Key    string `avro:"key"      json:"key,omitempty"`
    Value  string `avro:"value"    json:"value,omitempty"`
}
```